tests/e2e/raw_block_volume.go

```diff
@@ -1029,4 +1053,241 @@ var _ = ginkgo.Describe("raw block volume support", func() {
+			corev1.ReadWriteOnce, volumeSnapshot.Name, snapshotapigroup)
+		restorePvcSpec.Spec.VolumeMode = &rawBlockVolumeMode
+		restoredPvc, err := fpv.CreatePVC(client, namespace, restorePvcSpec)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		restoredPvs, err := fpv.WaitForPVClaimBoundPhase(client,
+			[]*corev1.PersistentVolumeClaim{restoredPvc},
+			framework.ClaimProvisionTimeout)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		volumeID2 := restoredPvs[0].Spec.CSI.VolumeHandle
+		gomega.Expect(volumeID2).NotTo(gomega.BeEmpty())
+
+		defer func() {
+			err := fpv.DeletePersistentVolumeClaim(client, restoredPvc.Name, namespace)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			err = e2eVSphere.waitForCNSVolumeToBeDeleted(volumeID2)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}()
+
+		// Create a Pod to use this PVC, and verify volume has been attached
+		ginkgo.By("Creating pod to attach restored PVC to the node")
+		pod2, err := createPod(client, namespace, nil, []*corev1.PersistentVolumeClaim{restoredPvc}, false,
+			"")
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		defer func() {
+			// Delete POD
+			ginkgo.By(fmt.Sprintf("Deleting the pod %s in namespace %s", pod2.Name, namespace))
+			err = fpod.DeletePodWithWait(client, pod2)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}()
+
+		nodeName = pod2.Spec.NodeName
+		vmUUID = getNodeUUID(ctx, client, pod2.Spec.NodeName)
+		ginkgo.By(fmt.Sprintf("Verify volume: %s is attached to the node: %s", volumeID2, nodeName))
+		isDiskAttached, err = e2eVSphere.isVolumeAttachedToVM(client, volumeID2, vmUUID)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		gomega.Expect(isDiskAttached).To(gomega.BeTrue(), "Volume is not attached to the node")
+
+		time.Sleep(5 * time.Minute)
+
+		// Verify previously written data. Later perform another write and verify it.
+		ginkgo.By(fmt.Sprintf("Verify previously written data on restored volume attached to: %v at path %v", pod2.Name,
+			pod2.Spec.Containers[0].VolumeDevices[0].DevicePath))
+		verifyDataFromRawBlockVolume(namespace, pod2.Name, devicePath, testdataFile, 0, 1)
+		ginkgo.By(fmt.Sprintf("Writing new 1mb test data in file %v", testdataFile))
+		op, err = exec.Command("dd", "if=/dev/urandom", fmt.Sprintf("of=%v", testdataFile),
+			"bs=1M", "count=1").Output()
+		fmt.Println(op)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		ginkgo.By(fmt.Sprintf("Write and read data on restored volume attached to: %v at path %v", pod2.Name,
+			pod2.Spec.Containers[0].VolumeDevices[0].DevicePath))
+		verifyIOOnRawBlockVolume(namespace, pod2.Name, devicePath, testdataFile, 0, 1)
+
+		// Cleanup
+		ginkgo.By("Delete volume snapshot and verify the snapshot content is deleted")
+		err = snapc.SnapshotV1().VolumeSnapshots(namespace).Delete(ctx, volumeSnapshot.Name, metav1.DeleteOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		snapshotCreated = false
```

Comment:
It is better to handle all cleanup operations in defer block  or add a check for this clean up task in After Each also. 
What if the code fails in deleting volume snapshot , The remaining code will not execute .

Reply:
done